# Data inspection - Check rendundacies and make the dataframe slimmer


In [1]:
import pandas as pd

In [2]:
df = pd.read_csv ('../data/accidents.csv', sep=';')

In [3]:
# First look at the dataframe
df.head(3)

IdUsager        Date     PV  Arrondissement                Mode  \
0   2389401  2017-04-03   3527           75111              Piéton   
1   2388322  2017-08-28   9113           75108  2 Roues Motorisées   
2   2394191  2017-11-06  11991           75117  2 Roues Motorisées   

    Catégorie       Gravité   Age     Genre  Milieu  ... Blessés Légers  \
0      Piéton  Blessé léger  62.0   Feminin  En-Agg  ...            1.0   
1  Conducteur  Blessé léger  30.0  Masculin  En-Agg  ...            1.0   
2  Conducteur  Blessé léger  37.0  Masculin  En-Agg  ...            1.0   

  Blessés hospitalisés Tué                                             Résumé  \
0                  NaN NaN  Accident Léger non mortel, En agglomération, H...   
1                  NaN NaN  Accident Léger non mortel, En agglomération, E...   
2                  NaN NaN  Accident Léger non mortel, En agglomération, E...   

        Coordonnées        Nom arrondissement arronco  \
0   48.855, 2.36867   Paris 4e Arrondissement   75104   
1   48.8667, 2.3013   Paris 8e Arrondissement   75108   
2  48.8858, 2.32163  Paris 17e Arrondissement   75117   

                                           arrondgeo     Coordonnées.1  \
0  {"coordinates": [[[[2.369123881, 48.853166231]...   48.855, 2.36867   
1  {"coordinates": [[[[2.301737288, 48.863496077]...   48.8667, 2.3013   
2  {"coordinates": [[[[2.303774362, 48.894153779]...  48.8858, 2.32163   

       Nom arrondissement.1  
0   Paris 4e Arrondissement  
1   Paris 8e Arrondissement  
2  Paris 17e Arrondissement  

[3 rows x 27 columns]

In [4]:
# Rename columns to English-friendly names for our beloved readers

df = df.rename(columns={
    'IdUsager': 'UserID',
    'Date': 'Date',
    'PV': 'ReportNumber',
    'Mode': 'TransportMode',
    'Catégorie': 'Category',
    'Gravité': 'Severity',
    'Age': 'Age',
    'Genre': 'Gender',
    'Milieu': 'Environment',
    'Adresse': 'Address',
    'Id accident': 'AccidentID',
    'PIM/BD PERIPHERIQUE': 'PeripheryInfo',
    "Tranche d'age": 'AgeGroup',
    'Blessés Légers': 'MinorInjuries',
    'Blessés hospitalisés': 'HospitalizedInjuries',
    'Tué': 'Fatalities',
    'Résumé': 'Summary',
    'Nom arrondissement': 'DistrictName',           
    'Nom arrondissement.1': 'DistrictName.1',
    'Coordonnées': 'Coordinates',
    'Coordonnées.1': 'Coordinates.1',
    'Arrondissement': 'District',
    'arronco': 'District_code'
})

df.head(3)


UserID        Date  ReportNumber  District       TransportMode  \
0  2389401  2017-04-03          3527     75111              Piéton   
1  2388322  2017-08-28          9113     75108  2 Roues Motorisées   
2  2394191  2017-11-06         11991     75117  2 Roues Motorisées   

     Category      Severity   Age    Gender Environment  ... MinorInjuries  \
0      Piéton  Blessé léger  62.0   Feminin      En-Agg  ...           1.0   
1  Conducteur  Blessé léger  30.0  Masculin      En-Agg  ...           1.0   
2  Conducteur  Blessé léger  37.0  Masculin      En-Agg  ...           1.0   

  HospitalizedInjuries Fatalities  \
0                  NaN        NaN   
1                  NaN        NaN   
2                  NaN        NaN   

                                             Summary       Coordinates  \
0  Accident Léger non mortel, En agglomération, H...   48.855, 2.36867   
1  Accident Léger non mortel, En agglomération, E...   48.8667, 2.3013   
2  Accident Léger non mortel, En agglomération, E...  48.8858, 2.32163   

               DistrictName District_code  \
0   Paris 4e Arrondissement         75104   
1   Paris 8e Arrondissement         75108   
2  Paris 17e Arrondissement         75117   

                                           arrondgeo     Coordinates.1  \
0  {"coordinates": [[[[2.369123881, 48.853166231]...   48.855, 2.36867   
1  {"coordinates": [[[[2.301737288, 48.863496077]...   48.8667, 2.3013   
2  {"coordinates": [[[[2.303774362, 48.894153779]...  48.8858, 2.32163   

             DistrictName.1  
0   Paris 4e Arrondissement  
1   Paris 8e Arrondissement  
2  Paris 17e Arrondissement  

[3 rows x 27 columns]

Let's create a summary of each column, including data types, missing values, and unique values. This helps us understand the data's structure and identify potential issues.

In [5]:
def column_summary(df):
    summary_data = []
    
    for col_name in df.columns:
        col_dtype = df[col_name].dtype
        num_of_nulls = df[col_name].isnull().sum()
        num_of_non_nulls = df[col_name].notnull().sum()
        num_of_distinct_values = df[col_name].nunique()
        
        summary_data.append({
            'col_name': col_name,
            'col_dtype': col_dtype,
            'num_of_nulls': num_of_nulls,
            'num_of_non_nulls': num_of_non_nulls,
            'num_of_distinct_values': num_of_distinct_values,
        })
    
    summary_df = pd.DataFrame(summary_data)
    return summary_df

summary_df = column_summary(df)
display(summary_df)

col_name col_dtype  num_of_nulls  num_of_non_nulls  \
0                 UserID     int64             0             41211   
1                   Date    object             0             41211   
2           ReportNumber     int64             0             41211   
3               District     int64             0             41211   
4          TransportMode    object             0             41211   
5               Category    object             0             41211   
6               Severity    object             0             41211   
7                    Age   float64             0             41211   
8                 Gender    object             0             41211   
9            Environment    object             0             41211   
10               Address    object             2             41209   
11               Champ13    object             0             41211   
12             Longitude    object             0             41211   
13              Latitude    object             0             41211   
14            AccidentID     int64             0             41211   
15         PeripheryInfo    object             0             41211   
16              AgeGroup    object           130             41081   
17         MinorInjuries   float64          3139             38072   
18  HospitalizedInjuries   float64         38294              2917   
19            Fatalities   float64         40957               254   
20               Summary    object          1016             40195   
21           Coordinates    object          5451             35760   
22          DistrictName    object          5451             35760   
23         District_code     int64             0             41211   
24             arrondgeo    object             0             41211   
25         Coordinates.1    object             0             41211   
26        DistrictName.1    object             0             41211   

    num_of_distinct_values  
0                    41211  
1                     2550  
2                     7105  
3                       40  
4                        6  
5                        3  
6                        3  
7                      103  
8                        2  
9                        2  
10                    4443  
11                      20  
12                   27647  
13                   23273  
14                   36156  
15                       2  
16                       9  
17                       1  
18                       1  
19                       1  
20                   26688  
21                   26906  
22                      43  
23                      44  
24                      44  
25                   31110  
26                      44

This summary suggests that some columns might be duplicates (e.g., DistrictName and DistrictName.1, Coordinates and Coordinates.1). We'll investigate these further.

In [6]:
district_mismatch = df[
    df['DistrictName'].notna() &
    (df['DistrictName'] != df['DistrictName.1'])
]
district_mismatch

Empty DataFrame
Columns: [UserID, Date, ReportNumber, District, TransportMode, Category, Severity, Age, Gender, Environment, Address, Champ13, Longitude, Latitude, AccidentID, PeripheryInfo, AgeGroup, MinorInjuries, HospitalizedInjuries, Fatalities, Summary, Coordinates, DistrictName, District_code, arrondgeo, Coordinates.1, DistrictName.1]
Index: []

[0 rows x 27 columns]

The check shows no differences, indicating that both `DistrictName` columns are redundant. I then proceed to verify whether the `Coordinates` and `Coordinates.1` columns, as well as the `Longitude` and `Latitude` columns, also contain exactly the same information. This consistency is important to confirm before dropping duplicates. Below is a simple equality check to see if any row has mismatched coordinate data. If it yields an empty result, the columns match perfectly.


In [7]:
coord_mismatch = df[
    df['Coordinates'].notna() &
    (df['Coordinates'] != df['Coordinates.1'])
]
coord_mismatch

latlong_mismatch = df[
    (df['Longitude'].notna()) & 
    (df['Latitude'].notna()) &
    (df['Coordinates'].notna()) &
    (df.apply(lambda row: f"{row['Latitude']}, {row['Longitude']}", axis=1) != df['Coordinates'])
]
latlong_mismatch

Empty DataFrame
Columns: [UserID, Date, ReportNumber, District, TransportMode, Category, Severity, Age, Gender, Environment, Address, Champ13, Longitude, Latitude, AccidentID, PeripheryInfo, AgeGroup, MinorInjuries, HospitalizedInjuries, Fatalities, Summary, Coordinates, DistrictName, District_code, arrondgeo, Coordinates.1, DistrictName.1]
Index: []

[0 rows x 27 columns]

Both resulting DataFrames (coord_mismatch and latlong_mismatch) are empty, I am confident that `Coordinates`, `Coordinates.1`, `Longitude`, and `Latitude` all store the same location details.


### Arrondissement data quality check

From the summary step, we also notice other arrondissement-related columns, such as `District` and `arronco`, which might differ slightly. Often, `DistrictName` (probably police data) and `arronco` (probably geocoding-based) can conflict near arrondissement boundaries.

To visually confirm the boundary issue, I opened Google Maps at the coordinates of an accident with conflicting district data (48.855, 2.3687). 
![alt text](image-1.png)

I see that the accident occurred right where the 4th and 11th districts meet. The police data labeled it as the 11th, whereas the geocoding approach correctly placed it in the 4th.

**This validates my decision to keep only one reliable district-related column in the final dataset. I choose `arronco` because it is derived from geolocation and typically has fewer missing values, thus more precise.**

Below, I remove columns that store data I no longer need (e.g., duplicates) or less accurate versions. I also convert remaining columns to their adequate data types and save the dataframe to a new CSV file.

In [8]:
columns_to_drop = [
    'DistrictName',     # Redundant if we keep geocoding-based arronco
    'Coordinates',      # Redundant if we rely on Latitude/Longitude
    'DistrictName.1',   # Also redundant
    'District',         # Less reliable than "arronco"
    'Coordinates.1',    # Also redundant
    'Champ13',          # Unused
    'arrondgeo'         # Polygon boundaries (not needed here)
]

df = df.drop(columns=columns_to_drop)

# Convert 'Date' column to datetime
df['Date'] = pd.to_datetime(df['Date'], errors='coerce')

# List of numeric columns
numeric_cols = [
    'UserID', 'ReportNumber', 'Age', 'AccidentID', 
    'MinorInjuries', 'HospitalizedInjuries', 'Fatalities'
]

# Convert numeric columns to nullable integer types
df[numeric_cols] = df[numeric_cols].apply(pd.to_numeric, errors='coerce').astype('Int64')

# List of string columns
string_cols = [
    'TransportMode', 'Category', 'Severity', 'Gender', 'Environment', 
    'Address', 'PeripheryInfo', 'AgeGroup', 'Summary'
]

# Convert string columns to the Pandas string dtype
df[string_cols] = df[string_cols].astype('string')

# Save the new df
df.to_csv('../data/accidents_cleaned.csv',index=False, sep=';' )

## Data preparation - Let's look closer at the `Summary` column.

Looking at rows details, we seem to have a enough information for an analysis of Paris traffic accidents. If we want to go deeper, it appears that the `Summary` column often offers a short, free-text narrative about how each accident occurred.

Now, we're embarking on the data parsing phase of our project. This is where we take the raw, messy text summaries of traffic accidents and transform them into structured data that we can analyze. You might think this would be straightforward, but it's surprisingly tricky.

Here's the problem: These summaries are full of incomplete phrases, like in this example: "Minor accident, Non-fatal, In urban area, T-intersection, Daylight, with Normal weather and Normal road surface. 1 Passenger vehicle (PV) traveling on Municipal Road (MR) driven by 1 Male user, 30 years old (Ind) hits 1 Veh". What does "Veh" stand for? Or consider this one: "Minor accident, Non-fatal, In urban area, Not at an intersection, Daylight, with Normal weather and Normal road surface. 1 Bicycle traveling on Municipal Road (MR) ridden by 1 Male user, 19 years old (Ind) hits 1 Pedestrian Male d". What comes after "d"? We can guess, but we need to be certain. It's not just about filling in the blanks; the structure of the summaries varies, and they use a lot of specialized abbreviations, like "VMA" for the maximum speed limit or "EDP-m" for a motorized personal mobility device.

We're not going to use fancy generative AI models for this task, though. Why? Because we need absolute accuracy and control. These AI models are great at generating text that looks right, but they can make mistakes, and we can't afford that when dealing with data that we'll be using for analysis. Plus, we need to understand exactly why a correction was made, and these models are like black boxes – it's hard to know what's going on inside. We'll use a more transparent and reliable approach, combining carefully crafted rules with some clever techniques to handle these tricky text snippets.